src/main/java/spliterators/part1/exercise/RectangleSpliterator.java

```diff
@@ -29,18 +29,36 @@ private RectangleSpliterator(int[][] array, int startOuterInclusive, int endOute
 
     @Override
     public OfInt trySplit() {
-        // TODO
-        throw new UnsupportedOperationException();
+        int length = endOuterExclusive - startOuterInclusive;
+        if (length <= 1)
+            return null;
+
+        int middle =  startOuterInclusive + length / 2;
+        RectangleSpliterator spliterator = new RectangleSpliterator(array, startOuterInclusive,
+                middle, 0);
+        startOuterInclusive = middle;
+        return spliterator;
     }
 
     @Override
     public long estimateSize() {
-        return ((long) endOuterExclusive - startOuterInclusive)*innerLength - startInnerInclusive;
+        return ((long) endOuterExclusive - startOuterInclusive) * innerLength - startInnerInclusive;
     }
 
     @Override
     public boolean tryAdvance(IntConsumer action) {
-        // TODO
-        throw new UnsupportedOperationException();
+        if (startInnerInclusive < array[startOuterInclusive].length) {
+            action.accept(array[startOuterInclusive][startInnerInclusive]);
+            startInnerInclusive ++;
+            return true;
+        } else if (startOuterInclusive < endOuterExclusive-1) {

```

Comment:
You can avoid code duplication by moving this check to the beginning of the method.